### PR TITLE
Migrate parts of `rustc_expand` to session diagnostics

### PR DESCRIPTION
**Pull request by Noratrieb**
This migrates everything but the `mbe` and `proc_macro` modules. It also contains a few cleanups and drive-by/accidental diagnostic improvements which can be seen in the diff for the UI tests.

### CODE REVIEW

File: compiler/rustc_expand/src/expand.rs [hunk 1]
```diff
@@ -939,38 +938,32 @@ pub fn parse_ast_fragment<'a>(
...
             }
-            _ => {}
-        }
-        err.emit();
+            _ => None,
+        };
+
+        parser.sess.emit_err(IncompleteParse {
+            span: def_site_span,
+            token: token.to_string(),
```
Comment (klensy) [inline compiler/rustc_expand/src/expand.rs:961]: This is Cow, so:
```suggestion
            token: token.into_owned(),
```
but given that this is error path, doesn't matter.

File: compiler/rustc_expand/src/expand.rs [hunk 2]
```diff
@@ -606,29 +610,26 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
...
-        let msg = format!(
-            "non-{kind} macro in {kind} position: {path}",
-            kind = kind.name(),
-            path = pprust::path_to_string(&mac.path),
-        );
-        self.cx.span_err(span, &msg);
+        self.cx.emit_err(WrongFragmentKind {
+            span,
+            kind: kind.name(),
+            name: pprust::path_to_string(&mac.path),
```
Comment (davidtwco) [inline compiler/rustc_expand/src/expand.rs:630]: I've suggested this somewhere else, but can we implement `IntoDiagnosticArg` for the path here and then just use the path directly in the diagnostic type?

File: compiler/rustc_expand/src/expand.rs [hunk 3]
```diff
@@ -606,29 +610,26 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
...
-                "consider increasing the recursion limit by adding a \
-                 `#![recursion_limit = \"{}\"]` attribute to your crate (`{}`)",
-                suggested_limit, self.cx.ecfg.crate_name,
-            ))
-            .emit();
+
+        self.cx.emit_err(RecursionLimitReached {
+            span: expn_data.call_site,
+            descr: expn_data.kind.descr(),
+            suggested_limit: suggested_limit.to_string(),
```
Comment (davidtwco) [inline compiler/rustc_expand/src/expand.rs:617]: `Limit` could probably implement `IntoDiagnosticArg`

File: compiler/rustc_expand/src/expand.rs [hunk 4]
```diff
@@ -939,38 +938,32 @@ pub fn parse_ast_fragment<'a>(
...
             }
-            _ => {}
-        }
-        err.emit();
+            _ => None,
+        };
+
+        parser.sess.emit_err(IncompleteParse {
+            span: def_site_span,
+            token: token.into_owned(),
```
Comment (davidtwco) [inline compiler/rustc_expand/src/expand.rs:961]: `Token` might be able to implement `IntoDiagnosticArg` too

File: compiler/rustc_expand/src/config.rs [hunk 5]
```diff
@@ -117,34 +114,34 @@ fn get_features(
...
             };
 
             if let Some(edition) = edition_enabled_features.get(&name) {
-                let msg =
-                    &format!("the feature `{}` is included in the Rust {} edition", name, edition);
-                span_handler.struct_span_warn_with_code(mi.span(), msg, error_code!(E0705)).emit();
+                sess.emit_warning(FeatureIncludedInEdition {
+                    span: mi.span(),
+                    feature: name,
+                    edition: edition.to_string(),
```
Comment (davidtwco) [inline compiler/rustc_expand/src/config.rs:143]: Can whatever type `edition` is implement `IntoDiagnosticArg`?

**Review by davidtwco — COMMENTED** [review]
Apologies for the delay in reviewing, this largely looks good to me, left a few comments.

**Comment by Nilstrieb** [comment]
Some of your suggested impls were already present and some had to be added.

**Review by davidtwco — APPROVED** [review]
Apologies for the delay in getting back to this, r=me after rebasing

**Pull request merged**